src/common/io.hpp

```diff
@@ -271,6 +271,47 @@ class StringReader : virtual public Reader {
 
 using Vsize = vector<uint8_t>::size_type;
 
+class ByteReader : virtual public Reader {
+private:
+	shared_ptr<vector<uint8_t>> emitter_;
+	Vsize bytes_read_ {0};
+
+public:
+	ByteReader(vector<uint8_t> &emitter) :
+		emitter_ {&emitter, [](vector<uint8_t> *vec) {}} {
+	}
+
+	ByteReader(shared_ptr<vector<uint8_t>> emitter) :
+		emitter_ {emitter} {
+	}
+
+	ExpectedSize Read(vector<uint8_t>::iterator start, vector<uint8_t>::iterator end) override;
+};
+
+class BufferedReader : virtual public io::Reader {
+private:
+	io::Reader &wrapped_reader_;
+	bool rewind_done_ {false};
+	bool stop_done_ {false};
+
+protected:
+	vector<uint8_t> buffer_;
+	io::ByteReader buffer_reader_;
+
+public:
+	BufferedReader(io::Reader &reader) :
+		wrapped_reader_ {reader},
+		buffer_reader_ {buffer_} {};
+
+	expected::ExpectedSize Read(
+		vector<uint8_t>::iterator start, vector<uint8_t>::iterator end) override;
+
+	error::Error Rewind();
+
+	void StopBuffering();
```

Comment:
@kacf This is my naive approach: save all the data and rewind.

However as it is designed now if the user does not call `StopBuffering()` we will copy the whole Artifact in memory :boom:  - I need to rethink a bit how to best limit it but please give me your preliminary feedback :pray: